fedspendingtransparency/usaspending-website
# [DEV-3004] Simplifying GlobalConstants config to rely on env variables and secret global constants file

# DNM until [this pr](https://github.com/fedspendingtransparency/usaspending-config/pull/503) is merged

## Possible Promotion Strategies
### DEV/STAGING
1. Merge [non-prod configuration changes with new globalConstants file](https://github.com/fedspendingtransparency/usaspending-config/pull/503) and put this branch in sandbox; if smoke testing passes, promote to DEV & Staging.
2. Merge [non-prod configuration changes with new globalConstants file](https://github.com/fedspendingtransparency/usaspending-config/pull/503), then merge, smoke test in DEV, then promote to Staging.
### PROD
After completing either  `1` or `2` above, merge [the production configuration changes](https://github.com/fedspendingtransparency/usaspending-config/pull/504) and promote to production.

<hr>
**High level description:**
We want two build configurations:

1. For the public to have access to the public api by default or their local instance of the api
2. For developers to use the lower envs locally and to successfully build for deployments to lower envs

**Technical details:**
`(1)`: Provide a `GlobalConstants.js` file in `src/js`
`(2)`: Expect secretGlobalConstant file at project root that exports correct config object based on env variable.

Implementation: 
Using `webpack.definePlugin()` to give access to env variable set by new docker script. Changed deployment script in [this pr](https://github.com/fedspendingtransparency/usaspending-config/pull/503) to use [this global-config file](https://github.com/fedspendingtransparency/usaspending-config/compare/fe_single_global_constants_file#diff-ec8b292d89552b6812190df0ccb6f763)

**JIRA Ticket:**
[DEV-3004](https://federal-spending-transparency.atlassian.net/browse/DEV-3004)

**Testing Changes in Docker**
These commands are taken straight from our config repos deployment logic:
1. Build Docker image at root of front end directory: `docker build --no-cache . -t `whoami`-usaspending-frontend`
2. Run Docker container: `docker run -i --rm=true --env GLOBAL_CONSTANTS_ENV='sandbox/development/production/staging' -v $(pwd)/public:/node-workspace/public `whoami`-usaspending-frontend /bin/sh -c 'npm run dev'
    docker run -p 8020:80 -v $(pwd)/public:/usr/share/nginx/html:ro nginx`
3. Go to `localhost:8020`
4. Verify base url is correct and import of `GlobalConstants/kGlobalConstants` succeed

The following are ALL required for the PR to be merged:
- [ ] Docker test above passes
- [ ] Send write up to relevant parties around setting up local env
